pkg/cmd/utils/gen-release-notes/generate.go

```diff
@@ -0,0 +1,183 @@
+			if err := o.Validate(); err != nil {
+				return err
+			}
+
+			ghClient := githubClient(ctx)
+			pullRequests, err := pullRequestsFromMergeCommits(ctx, ghClient, o)
+			if err != nil {
+				return err
+			}
+
+			if err := renderReleaseNotes(o.Out, o.Release, pullRequests); err != nil {
+				return err
+			}
+
+			return nil
+		},
+	}
+
+	cmd.Flags().StringVar(&o.Release, "release", o.Release, "Name of the release.")
+	cmd.Flags().StringVar(&o.RepositoryPath, "repository-path", o.RepositoryPath, "Path to git repository.")
+	cmd.Flags().StringVar(&o.TagFrom, "tag-from", o.TagFrom, "PRs merged after before this tag won't be considered.")
+	cmd.Flags().StringVar(&o.TagTo, "tag-to", o.TagTo, "PRs merged before before this tag won't be considered.")
+
+	return cmd
+}
+
+func pullRequestsFromMergeCommits(ctx context.Context, ghClient *github.Client, o *GenerateOptions) ([]*github.PullRequest, error) {
+	repo, err := git.PlainOpen(o.RepositoryPath)
+	if err != nil {
+		return nil, err
+	}
+
+	commits, err := listCommitsBetween(repo, o.TagFrom, o.TagTo)
+	if err != nil {
+		return nil, err
+	}
+
+	pullRequests := make([]*github.PullRequest, 0, len(commits))
+	for _, c := range commits {
+		if !isMergeCommit(c) {
+			continue
+		}
+
+		n, err := parsePullRequestNumber(c.Message)
+		if err != nil {
+			if errors.Is(ErrNotAPullRequest, err) {
+				continue
+			}
+			return nil, err
+		}
+
+		pr, _, err := ghClient.PullRequests.Get(ctx, o.OrganizationName, o.RepositoryName, n)
+		if err != nil {
+			return nil, err
+		}
+		pullRequests = append(pullRequests, pr)
+	}
+
+	return pullRequests, nil
+}
+
+func isMergeCommit(commit *object.Commit) bool {
+	return commit.NumParents() > 1
+}
+
+var (
+	pullRequestNumberRe = regexp.MustCompile(`^Merge pull request #(\d+) .*$`)
+	ErrNotAPullRequest  = fmt.Errorf("commit is not a pull request")
+)
+
+// Merge commit originating from merged PR has following format:
+// 'Merge pull request #<pr-number> from <author>/<branch>'
+func parsePullRequestNumber(message string) (int, error) {
+	firstLine := strings.Split(message, "\n")[0]
+	m := pullRequestNumberRe.FindStringSubmatch(firstLine)
+	if m == nil {
+		return 0, ErrNotAPullRequest
+	}
+
+	return strconv.Atoi(m[1])
+}
+
+func commitFromTag(r *git.Repository, tag string) (*object.Commit, error) {
+	ref, err := r.Tag(tag)
+	if err != nil {
+		return nil, fmt.Errorf("tag %w", err)
+	}
+
+	t, err := r.TagObject(ref.Hash())
+	if err != nil {
+		return nil, fmt.Errorf("tag object, %w", err)
+	}
+
+	commit, err := r.CommitObject(t.Target)
+	if err != nil {
+		return nil, fmt.Errorf("commit object, %w", err)
+	}
+
+	return commit, nil
+}
+
+func listCommitsBetween(r *git.Repository, tagFrom, tagTo string) ([]*object.Commit, error) {
+	fromCommit, err := commitFromTag(r, tagFrom)
+	if err != nil {
+		return nil, err
+	}
+
+	toCommit, err := commitFromTag(r, tagTo)
+	if err != nil {
+		return nil, err
+	}
+
+	mergeBaseCommits, err := toCommit.MergeBase(fromCommit)
```

Comment:
Doesn't the start commit already have to me a merge-base? I was assuming the start commit is already a direct predecessor of the end commit, otherwise the PR diff would be 2 ways. 

- for diff between `v1.1.0` and `v1.1.1` this is fine
- when we want a diff between `v1.2.0` and the previous release base, I think we'd rather find the merge-base externally

Reply:
changed to sanity check